sycl/doc/extensions/proposed/sycl_ext_intel_event_mode.asciidoc

```diff
@@ -0,0 +1,165 @@
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 9 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+This extension also depends on the following other SYCL extensions:
+
+* link:../experimental/sycl_ext_oneapi_enqueue_functions.asciidoc[
+  sycl_ext_oneapi_enqueue_functions]
+* link:../experimental/sycl_ext_oneapi_properties.asciidoc[
+  sycl_ext_oneapi_properties]
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+On some backends, calling `wait()` on an `event` will synchronize using a
+busy-waiting implementation. Though this comes at a low latency for the
+synchronization of the event, it has the downside of consuming high amounts of
+CPU time for no meaningful work.  This extension introduces a new property for
+SYCL commands that allow users to pick modes for the associated events, one of
+these modes being a "low-power" event. These new low-power events will, if
+possible, yield the thread that the `wait()` member function is called on and
+only wake up occasionally to check if the event has finished. This reduces the
+time the CPU spends checking finish condition of the wait, at the cost of
+latency.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_INTEL_EVENT_MODE` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Event mode property
+
+This extension adds a new property `event_mode` which can be used with the
+`submit_with_event` free function from
+link:../experimental/sycl_ext_oneapi_enqueue_functions.asciidoc[sycl_ext_oneapi_enqueue_functions],
+allowing the user some control over how the resulting event is created and
+managed.
+
+```
+namespace sycl::ext::intel::experimental {
+
+enum class event_modes { none, low_power };
```

Comment:
@Pennycook: Do you like this naming convention where the enum name is plural and the property name is singular?

The other alternative we discussed is to name the property `event_mode` and name the enum `event_mode_enum`.

Reply:
A quick search of the ISO C++ specification suggests that their convention is to use a plural when the enum is a bitmask (e.g., `std::filesystem::copy_options`) but a singular when it is not (e.g., `std::filesystem::file_type`).

If we want to keep the plural here, then I think we should define this as a bitmask.  But that also suggests that the property name should also be a plural.  Didn't we recently discuss reintroducing a nested namespace for properties?  That would allow `sycl::event_modes` to be a bitmask of event modes, and `sycl::property::event_modes` to be a property for setting an event's modes.

I don't really like the idea of encoding `_enum` in a name, because of the redundancy there.

Reply:
I don't think we should make this a bitmask.

We have already used the `_enum` suffix for some other properties, so we have a precedent.

If we did add the namespace, why wouldn't we put the property and the enum in the same namespace?  The enum is only used in conjunction with the property.

Reply:
@steffenlarsen, I saw that you re-requested my review.  I think the naming issue in this thread is not yet resolved.

If we can't think of a better solution, I suggest we go with the name `event_mode_enum`, which is consistent with existing properties that take an enum.

Reply:
Ah, apologies. I glanced over this one. `event_mode_enum` is fine with me. 👍